relayer/relays/beefy/on-demand-sync.go

```diff
@@ -72,101 +83,93 @@ func (relay *OnDemandRelay) Start(ctx context.Context) error {
+			if err != nil {
+				return fmt.Errorf("sync for v2 message: %w", err)
 			}
-			log.WithError(err).Error("Query nonces")
-			continue
 		}
+	}
+}
 
-		log.WithFields(log.Fields{
-			"paraNonce": paraNonce,
-			"ethNonce":  ethNonce,
-		}).Info("Nonces checked")
+func (relay *OnDemandRelay) syncV1(ctx context.Context) error {
+	paraNonce, ethNonce, err := relay.queryNonces(ctx)
+	if err != nil {
+		return fmt.Errorf("Query nonces: %w", err)
+	}
 
-		if paraNonce > ethNonce {
+	log.WithFields(log.Fields{
+		"paraNonce": paraNonce,
+		"ethNonce":  ethNonce,
+	}).Info("Nonces checked")
 
-			// Check if we are rate-limited
-			if !relay.tokenBucket.TryConsume(1) {
-				log.Info("Rate-limit exceeded")
-				continue
-			}
+	if paraNonce > ethNonce {
 
-			log.Info("Performing sync")
+		// Check if we are rate-limited
+		if !relay.tokenBucket.TryConsume(1) {
+			return fmt.Errorf("Rate-limit exceeded")
+		}
 
-			beefyBlockHash, err := relay.relaychainConn.API().RPC.Beefy.GetFinalizedHead()
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
-				log.WithError(err).Error("Fetch latest beefy block hash")
-				continue
-			}
+		log.Info("Performing sync v1")
 
-			header, err := relay.relaychainConn.API().RPC.Chain.GetHeader(beefyBlockHash)
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
-				log.WithError(err).Error("Fetch latest beefy block header")
-				continue
-			}
-
-			err = relay.sync(ctx, uint64(header.Number))
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
-				log.WithError(err).Error("Sync failed")
-				continue
-			}
+		beefyBlockHash, err := relay.relaychainConn.API().RPC.Beefy.GetFinalizedHead()
+		if err != nil {
+			return fmt.Errorf("Fetch latest beefy block hash: %w", err)
+		}
 
-			log.Info("Sync completed")
+		header, err := relay.relaychainConn.API().RPC.Chain.GetHeader(beefyBlockHash)
+		if err != nil {
+			return fmt.Errorf("Fetch latest beefy block header: %w", err)
+		}
 
-			relay.waitUntilMessagesSynced(ctx, paraNonce)
+		err = relay.sync(ctx, uint64(header.Number))
+		if err != nil {
+			return fmt.Errorf("Sync failed: %w", err)
 		}
+
+		log.Info("Sync completed")
+
+		relay.waitUntilMessagesSynced(ctx, paraNonce)
 	}
+	return nil
 }
 
 func (relay *OnDemandRelay) waitUntilMessagesSynced(ctx context.Context, paraNonce uint64) {
-	sleep(ctx, time.Minute*10)
+	waitingTime := 0
 	for {
 		ethNonce, err := relay.fetchEthereumNonce(ctx)
 		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				return
-			}
 			log.WithError(err).Error("fetch latest ethereum nonce")
-			sleep(ctx, time.Minute*1)
-			continue
+			return
 		}
-
-		if ethNonce >= paraNonce {
+		waitingTime++
+		if ethNonce >= paraNonce || waitingTime > 10 {
 			return
 		}
-	}
-
-}
-
-func sleep(ctx context.Context, d time.Duration) {
-	select {
-	case <-ctx.Done():
-		return
-	case <-time.After(d):
+		time.Sleep(time.Second * 10)
```

Comment:
Why change the implementation of `sleep`? This new implementation doesn't handle the case when `ctx` is cancelled, and it means that the relayer process won't exit properly on error.

Reply:
> This new implementation doesn't handle the case when ctx is cancelled, and it means that the relayer process won't exit properly on error.

This won't happen, as we will run the on-demand relay in a separate goroutine. As mentioned in https://github.com/Snowfork/snowbridge/pull/1449#discussion_r2079728457.

That way, we can remove the internal checks everywhere and make error handling cleaner.